dialer_test.go

```diff
@@ -0,0 +1,155 @@
+package cloudsqlconn
+
+import (
+	"context"
+	"io/ioutil"
+	"net/http"
+	"strings"
+	"testing"
+
+	"cloud.google.com/cloudsqlconn/internal/mock"
+	"google.golang.org/api/option"
+	sqladmin "google.golang.org/api/sqladmin/v1beta4"
+)
+
+// replaceSQLAdminClient patches the Dialer's SQL Admin API client with one that
+// uses the provided HTTP client and endpoint.
+func replaceSQLAdminService(d *Dialer, client *http.Client, endpoint string) {
+	svc, err := sqladmin.NewService(
+		context.Background(),
+		option.WithHTTPClient(client),
+		option.WithEndpoint(endpoint),
+	)
+	if err != nil {
+		panic(err)
+	}
+	d.sqladmin = svc
+}
+
+func TestDialerCanConnectToInstance(t *testing.T) {
+	inst := mock.NewFakeCSQLInstance("my-project", "my-region", "my-instance")
+	mc, url, cleanup := mock.HTTPClient(
+		mock.InstanceGetSuccess(inst, 1),
+		mock.CreateEphemeralSuccess(inst, 1),
+	)
+	stop := mock.StartServerProxy(mock.ServerProxyConfig{
+		Response: "server-response",
+		Instance: inst,
+	})
+	defer func() {
+		stop()
+		if err := cleanup(); err != nil {
+			t.Fatalf("%v", err)
+		}
+	}()
+
+	d, err := NewDialer(context.Background(), WithDefaultDialOptions(WithPublicIP()))
+	if err != nil {
+		t.Fatalf("expected NewDialer to succeed, but got error: %v", err)
+	}
+	replaceSQLAdminService(d, mc, url)
```

Comment:
I'd rather have this be more explicit that's happening - it seems like `d.sqladmin := SQLAdminService...` is more clear that the dialer is being mutated.

Reply:
I agree. That reads better.